src/hotspot/share/opto/superword.cpp

```diff
@@ -3807,3 +3036,412 @@ bool SuperWord::same_origin_idx(Node* a, Node* b) const {
+    shift_count->set_req(1, unique_vtn);
+    return shift_count;
+  }
+
+  if (unique != nullptr) {
+    // Replicate the scalar unique to every vector element. The resulting elements must
+    // be of the same velt_type as p0.
+    VTransformNode* unique_vtn = find_scalar(unique);
+    const Type* element_type = _vloop_analyzer.types().velt_type(p0);
+
+    // Scalar rotate has int rotation value. But if we are rotating longs, then we must
+    // convert the int rotation to long.
+    if (j == 2 && VectorNode::is_scalar_rotate(p0) && element_type->isa_long()) {
+      assert(unique->bottom_type()->isa_int(), "scalar rotate expects int rotation");
+      VTransformNode* conv = new (_graph.arena()) VTransformConvI2LNode(_graph);
+      conv->set_req(1, unique_vtn);
+      unique_vtn = conv;
+    }
+
+    VTransformNode* replicate = new (_graph.arena()) VTransformReplicateNode(_graph, pack->size(), element_type);
+    replicate->set_req(1, unique_vtn);
+    return replicate;
+  }
+
+  // The input is neither a pack not a unique node. SuperWord::profitable does not allow
+  // any other case. In the future, we could insert a PackNode.
+#ifdef ASSERT
+  tty->print_cr("\nSuperWordVTransformBuilder::find_input_for_vector: j=%d", j);
+  pack->dump();
+  assert(false, "Pack input was neither a pack nor a unique node");
+#endif
+  ShouldNotReachHere();
+}
+
+VTransformNode* SuperWordVTransformBuilder::find_scalar(Node* n) {
+  VTransformNode* vtn = get_vtnode_or_null(n);
+  if (vtn != nullptr) { return vtn; }
+
+  assert(!_vloop.in_bb(n), "only nodes outside the loop can be input nodes to the loop");
+  vtn = new (_graph.arena()) VTransformInputScalarNode(_graph, n);
+  set_vtnode(n, vtn);
+  return vtn;
+}
+
+void SuperWordVTransformBuilder::set_req_for_vector(VTransformNode* vtn, VectorSet& vtn_dependencies, int j, Node_List* pack) {
+  VTransformNode* req = find_input_for_vector(j, pack);
+  vtn->set_req(j, req);
+  vtn_dependencies.set(req->_idx);
+}
+
+void SuperWordVTransformBuilder::set_req_all_for_scalar(VTransformNode* vtn, VectorSet& vtn_dependencies, Node* n) {
+  assert(vtn->req() == n->req(), "scalars must have same number of reqs");
+  for (uint j = 0; j < n->req(); j++) {
+    Node* def = n->in(j);
+    if (def == nullptr) { continue; }
+    set_req_for_scalar(vtn, vtn_dependencies, j, n);
+  }
+}
+
+void SuperWordVTransformBuilder::set_req_all_for_vector(VTransformNode* vtn, VectorSet& vtn_dependencies, Node_List* pack) {
+  Node* p0 = pack->at(0);
+  assert(vtn->req() <= p0->req(), "must have at at most as many reqs");
+  // Vectors have no ctrl, so ignore it.
+  for (uint j = 1; j < vtn->req(); j++) {
+    Node* def = p0->in(j);
+    if (def == nullptr) { continue; }
+    set_req_for_vector(vtn, vtn_dependencies, j, pack);
+  }
+}
+
+void SuperWordVTransformBuilder::add_dependencies(VTransformNode* vtn, VectorSet& vtn_dependencies, Node* n) {
```

Comment:
Note: see `PacksetGraph::build` for the old `VLoopDependencyGraph::PredsIterator` usage.